File: app/src/main/java/com/duzhaokun123/bilibilihd/utils/NotificationUtil.java
package com.duzhaokun123.bilibilihd.utils;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationChannelGroup;
import android.app.NotificationManager;
import android.content.Context;
import android.util.Log;

import com.duzhaokun123.bilibilihd.R;

import java.util.HashMap;
import java.util.Map;

public class NotificationUtil {
    public static String CHANNEL_GROUP_ID_DOWNLOAD = "download";
    public static String CHANNEL_ID_VIDEO_DOWNLOAD = "video_download";

    private static Map<Integer, Notification> notificationMap;

    public static boolean isIdUnregistered(int id) {
        return notificationMap == null || notificationMap.get(id) == null;
    }

    public static void init(Context context) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager == null) {
            return;
        }
        NotificationChannelGroup notificationChannelGroup = new NotificationChannelGroup(CHANNEL_GROUP_ID_DOWNLOAD, context.getString(R.string.download));
        notificationManager.createNotificationChannelGroup(notificationChannelGroup);

        NotificationChannel notificationChannel = new NotificationChannel(CHANNEL_ID_VIDEO_DOWNLOAD, context.getString(R.string.video_download), NotificationManager.IMPORTANCE_LOW);
        notificationChannel.enableLights(true);
        notificationChannel.enableVibration(true);
        notificationChannel.setGroup(CHANNEL_GROUP_ID_DOWNLOAD);
        notificationManager.createNotificationChannel(notificationChannel);
    }

    public static void show(Context context, int id, Notification notification) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager == null) {
            return;
        }
        if (notificationMap == null) {
            notificationMap = new HashMap<>();
        }
        if (notificationMap.get(id) != null) {
            Log.e("NotificationUtil", "should use reshow");
            return;
        }
        notificationMap.put(id, notification);
        notificationManager.notify(id, notification);
    }

    public static void reshow(Context context, int id) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager == null || notificationMap == null) {
            return;
        }
        if (notificationMap.get(id) == null) {
            return;
        }
        notificationManager.notify(id, notificationMap.get(id));
    }

    public static void reshow(Context context, int id, Notification notification) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager == null || notificationMap == null) {
            return;
        }
        if (notificationMap.get(id) == null) {
            return;
        }
        notificationMap.put(id, notification);
        notificationManager.notify(id, notification);
    }

    public static void remove(Context context, int id) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager == null || notificationMap == null) {
            return;
        }
        notificationManager.cancel(id);
        notificationMap.remove(id);
    }

    public static Notification getNotification(int id) {
        return notificationMap == null ? null : notificationMap.get(id);
    }

    public static void makeNotificationCleanable(Notification notification) {
        if (notification != null) {
            notification.flags &= ~ Notification.FLAG_NO_CLEAR;
        }
    }

    public static void unregister(int id) {
        if (notificationMap != null) {
            notificationMap.remove(id);
        }
    }
}

File: app/src/main/java/com/duzhaokun123/bilibilihd/utils/NotificationUtil.kt
package com.duzhaokun123.bilibilihd.utils

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationChannelGroup
import android.app.NotificationManager
import android.content.Context
import com.duzhaokun123.bilibilihd.R
import java.util.*

object NotificationUtil {
    const val CHANNEL_GROUP_ID_PLAY = "play"
    const val CHANNEL_ID_VIDEO_PLAY_BACKGROUND = "play_background"

    private var notificationMap: MutableMap<Int, Notification> = HashMap()
    private var context: Context? = null
    private var notificationManager: NotificationManager? = null

    fun isIdUnregistered(id: Int): Boolean {
        return notificationMap[id] == null
    }

    fun setContext(context: Context) {
        this.context = context
        this.notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    }

    fun getNewId(): Int {
        while (true) {
            val id: Int = System.currentTimeMillis().toInt()
            if (id != 0 && isIdUnregistered(id)) {
                return id
            }
        }
    }

    fun init(context: Context) {
        val notificationChannelGroup = NotificationChannelGroup(CHANNEL_GROUP_ID_PLAY, context.getString(R.string.play))
        notificationManager?.createNotificationChannelGroup(notificationChannelGroup)
        val notificationChannel = NotificationChannel(CHANNEL_ID_VIDEO_PLAY_BACKGROUND, context.getString(R.string.play_background), NotificationManager.IMPORTANCE_MIN)
        notificationChannel.enableLights(false)
        notificationChannel.enableVibration(true)
        notificationChannel.group = CHANNEL_GROUP_ID_PLAY
        notificationManager?.createNotificationChannel(notificationChannel)
    }

    fun show(id: Int, notification: Notification) {
        notificationMap[id] = notification
        try {
            notificationManager?.notify(id, notification)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun reshow(id: Int) {
        if (notificationMap[id] == null) {
            return
        }
        try {
            notificationManager?.notify(id, notificationMap[id])
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun reshow(id: Int, notification: Notification?) {
        if (notificationMap[id] == null || notification == null) {
            return
        }
        notificationMap[id] = notification
        try {
            notificationManager?.notify(id, notification)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun remove(id: Int) {
        notificationManager?.cancel(id)
        notificationMap.remove(id)
    }

    fun getNotification(id: Int): Notification? {
        return notificationMap[id]
    }

    /**
     * FIXME: 无法设置为可清除
     */
    fun setNotificationCleanable(id: Int, cleanable: Boolean) {
        val notification = getNotification(id)
        if (notification != null) {
            if (cleanable) {
                notification.flags = (notification.flags and Notification.FLAG_NO_CLEAR).inv()
            } else {
                notification.flags = notification.flags or Notification.FLAG_NO_CLEAR
            }
        }
        reshow(id)
    }

    fun unregister(id: Int) {
        notificationMap.remove(id)
    }
}